docs/design/gh-ost-integration.md

```diff
@@ -0,0 +1,509 @@
+
+We will divide gh-ost execution into different tasks. Each task needs to be approved by user to proceed. These tasks are described in detail in the following sections.
+
+1. Pre-migration check. We run gh-ost with noop flag to check user's input.
+2. In-migration,
+    1. After approval, gh-ost performs the migration, syncs the ghost table with the original table.
+    2. After the ghost table is synced with the original table, there is a cut-over phase. We will only allow gh-ost to do the cut-over if user approved.
+3. After migration, the final task is to delete the original table if user approved.
+
+### Pre-migration validation
+
+gh-ost has many requirements and limitations, we want to check whether gh-ost is good to go before the actual migration. To validate it, we run gh-ost with noop flag. We will return the error, if any, to user and let user handle it. This will be a new task check run type.
+
+```Go
+const TaskCheckDatabaseSchemaUpdateGhost = "bb.task-check.database.schema.update.ghost"
+type TaskCheckDatabaseSchemaUpdateGhostPayload struct {
+    Statement string `json:"statement,omitempty"`
+    InstanceID int `json:"instanceId,omitempty"`
+    DatabaseName string `json:"databaseName,omitempty"`
+    TableName string `json:"tableName,omitempty"`
+}
+```
+
+### In-migration progress
+
+gh-ost can report its migration progress. Migrator can write migration status to io.Writer. [Related codes here](https://github.com/github/gh-ost/blob/master/go/logic/migrator.go#L906).
+
+```Go
+// printStatus prints the progress status, and optionally additionally detailed
+// dump of configuration.
+// `rule` indicates the type of output expected.
+// By default the status is written to standard output, but other writers can
+// be used as well.
+func (this *Migrator) printStatus(rule PrintStatusRule, writers ...io.Writer) {
+}
+```
+
+It's important to show user the progress, because typically it will take hours or even days to complete when migrating on large tables. We periodically update the progress in `TaskRunResultPayload.Detail`.
+
+### Approve, run, approve, cut-over
+
+We will have two tasks for the actual migration. The first task runs gh-ost, syncs the ghost table with the original table. Once synced, the first task will be `DONE` and the second task will be `PENDING_APPROVAL`. The second task does the cut-over, which locks the tables and atomically switch table names. The second task is blocked by the first task.
+
+These two tasks need to access the same `Migrator`. Unfortunately, the way we schedule tasks now is not capable of it. Go channels to the rescue! For go routines that run tasks, we will store go channels globally to communicate with them. When the ghost table catches up with the original table , the "run" task will switch to `DONE`, and the "cut over" task will be `PENDING_APPROVAL`. The "cut over" task sends `"cut-over"` to the "run" task go routine to execute cut over.
+
+```Go
+type TaskScheduler struct {
```

Comment:
According to the recent offline discussion, this section needs to be refined. IIUC, there is no need to keep the map